### PR TITLE
aws_credentials_provider binding

### PR DESCRIPTION
**Pull request by graebm**
Only exposing "default chain" and "static" providers for now

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/auth.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+PyObject *aws_py_credentials_provider_get_credentials(PyObject *self, PyObject *args);
+PyObject *aws_py_credentials_provider_shutdown(PyObject *self, PyObject *args);
+
+PyObject *aws_py_credentials_provider_new_chain_default(PyObject *self, PyObject *args);
+PyObject *aws_py_credentials_provider_new_static(PyObject *self, PyObject *args);
+
+/* Given a python object, return a pointer to its underlying native type.
+ * If NULL is returned, a python error has been set */
+
+struct aws_credentials_provider *aws_py_get_credentials_provider(PyObject *credentials_provider);
```
Comment (ColdenCullen) [inline source/auth.h]: fuck I love this pattern so much

File: source/auth_credentials.c [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+/* Runs when the GC destroys the capsule containing the binding */
+static void s_credentials_provider_capsule_destructor(PyObject *capsule) {
+    struct credentials_provider_binding *provider = PyCapsule_GetPointer(capsule, s_capsule_name_credentials_provider);
+    if (provider->native) {
+        aws_credentials_provider_release(provider->native);
+    }
+
+    Py_XDECREF(provider->bootstrap);
+}
```
Comment (ColdenCullen) [inline source/auth_credentials.c]: does `provider` need to be released?

Comment (graebm) [reply source/auth_credentials.c]: GOOD F'ING CATCH

File: source/auth_credentials.c [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+int s_aws_string_to_cstr_and_ssize(const struct aws_string *source, const char **out_cstr, Py_ssize_t *out_ssize) {
+    *out_cstr = NULL;
+    *out_ssize = 0;
+    if (source) {
+        if (source->len > PY_SSIZE_T_MAX) {
+            return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+        }
+        *out_cstr = (const char *)source->bytes;
```
Comment (JonathanHenson) [inline source/auth_credentials.c]: don't we have functions for this?

Comment (graebm) [reply source/auth_credentials.c]: **LOOK WHAT YOU MADE ME DO**
https://github.com/awslabs/aws-c-common/pull/534

**Pull request merged**